test/e2e/lua/dynamic_certificates.go

```diff
@@ -0,0 +1,175 @@
+		Expect(len(errs)).Should(BeNumerically("==", 0))
+		Expect(resp.StatusCode).Should(Equal(http.StatusOK))
+
+		Expect(err).ToNot(HaveOccurred())
+	})
+
+	Context("when only servers change", func() {
+		It("should handle SSL certificate only changes", func() {
+			ingress, err := f.KubeClientSet.ExtensionsV1beta1().Ingresses(f.IngressController.Namespace).Get("foo.com", metav1.GetOptions{})
+			Expect(err).ToNot(HaveOccurred())
+
+			ingress.Spec.TLS = []extensions.IngressTLS{
+				{
+					Hosts:      []string{"foo.com"},
+					SecretName: "foo.com",
+				},
+			}
+
+			_, err = framework.CreateIngressTLSSecret(f.KubeClientSet,
+				ingress.Spec.TLS[0].Hosts,
+				ingress.Spec.TLS[0].SecretName,
+				ingress.Namespace)
+			Expect(err).ToNot(HaveOccurred())
+
+			_, err = f.KubeClientSet.ExtensionsV1beta1().Ingresses(f.IngressController.Namespace).Update(ingress)
+			Expect(err).ToNot(HaveOccurred())
+
+			err = f.WaitForNginxServer("foo.com",
+				func(server string) bool {
+					return strings.Contains(server, "server_name foo.com") &&
+						strings.Contains(server, "listen 443")
+				})
+			Expect(err).ToNot(HaveOccurred())
+
+			log, err := f.NginxLogs()
+			Expect(err).ToNot(HaveOccurred())
+			Expect(log).ToNot(BeEmpty())
+
+			By("POSTing new certificates to Lua endpoint")
+			Expect(log).To(ContainSubstring(logDynamicConfigSuccess))
+			Expect(log).ToNot(ContainSubstring(logDynamicConfigFailure))
+
+			By("skipping Nginx reload")
+			Expect(log).ToNot(ContainSubstring(logRequireBackendReload))
+			Expect(log).ToNot(ContainSubstring(logBackendReloadSuccess))
+			Expect(log).To(ContainSubstring(logSkipBackendReload))
+			Expect(log).ToNot(ContainSubstring(logInitialConfigSync))
+		})
+	})
+
+	Context("when certificates are requested", func() {
+		It("should serve certificates dynamically from Lua", func() {
+			ingress, err := f.KubeClientSet.ExtensionsV1beta1().Ingresses(f.IngressController.Namespace).Get("foo.com", metav1.GetOptions{})
+			Expect(err).ToNot(HaveOccurred())
+
+			ingress.Spec.TLS = []extensions.IngressTLS{
+				{
+					Hosts:      []string{"foo.com"},
+					SecretName: "foo.com",
+				},
+			}
+
+			_, err = framework.CreateIngressTLSSecret(f.KubeClientSet,
+				ingress.Spec.TLS[0].Hosts,
+				ingress.Spec.TLS[0].SecretName,
+				ingress.Namespace)
+			Expect(err).ToNot(HaveOccurred())
+
+			_, err = f.KubeClientSet.ExtensionsV1beta1().Ingresses(f.IngressController.Namespace).Update(ingress)
+			Expect(err).ToNot(HaveOccurred())
+
+			err = f.WaitForNginxServer("foo.com",
+				func(server string) bool {
+					return strings.Contains(server, "server_name foo.com") &&
+						strings.Contains(server, "listen 443")
+				})
+			Expect(err).ToNot(HaveOccurred())
+
+			By("checking SSL Certificate using the NGINX IP address")
+			resp, _, errs := gorequest.New().
+				Get(f.IngressController.HTTPSURL).
+				Set("Host", ingress.Spec.TLS[0].Hosts[0]).
+				TLSClientConfig(&tls.Config{
+					InsecureSkipVerify: true,
+				}).
```

Comment:
since you are providing a custom config you have to explicitly set the `ServerName` to `ingress.Spec.TLS[0].Hosts[0]`